[WIP] Virtual product grouped datasets

uchchwhash opened the pull request:
This is a request for review of the general structure of the code.
Just something concrete enough to refer back to.

Notes:
- except for `Collate` the other classes "work" at the moment
- essentially all the names are subject to bike-shedding so feel free
- please do not merge yet

uchchwhash committed:
time-stamped datasets; snafu

uchchwhash committed:
these changes break collate; will fix

mpaget commented:
A comment:
Could you provide some information on what a 'virtual-product' is? I don't see that this has been mentioned previously in any ODC (beyond GA) discussions, and there isn't much information in this pull request to support the 'why' or 'what problem it solves' questions. It would appear to border more on an enhancement proposal rather than a bug fix or efficiency improvement, so I would expect the proposal to be circulated to the wider ODC community for input. It may be appropriate to use the ODC developers meeting in February to discuss this further .. but difficult to assess that without any supporting information.

omad commented:
Matt, I'm not directly involved with this work, but it's very much in progress, hence the reason this pull request isn't against `develop`, but between two feature branches.

I'm sure that documentation will be forthcoming.

Kirill888 commented:
@mpaget proper design doc for this is coming, it's easier to document a working prototype. But to give an idea of what this all about here is a brief overview:

You can think of usual `datacube.load` as 2 step process

1. Build 3d VRT (virtual raster) for a given user query (space,time, product)
2. Load that VRT into memory at a requested projection/resolution resulting in `xarray.Dataset`

What we call "virtual products" generalise this to support

1. Mixing of multiple on-disk products (NBAR + PQ for example)
2. Applying user supplied data transforms at load time (e.g. (NBAR+PQ)->NBAR_clear_pixels_only)
3. Treating multiple similar product as one (e.g. NBAR from LS5,LS7,LS8)

Why?

- Increased memory/load time efficiency when dealing with multi-product workflows
  - No need to load data that will be discarded (e.g. missing PQ for a given NBAR tile)
  - Transforming data time slice by time slice reduces peak memory requirement compared to current approach of loading all source data followed by a transform.
- Better UX and code re-use. 
  - User will be able to just say `clear_pixel(product='ls5_nbar_albers')` rather manually loading two product and reconciling data post-load.

uchchwhash committed:
more notes

uchchwhash committed:
Datacube.load, GridWorkflow.load compatibility

uchchwhash committed:
progress; but fetch_data needs work

uchchwhash committed:
need more structure to grouped datasets

uchchwhash committed:
no time sorting for now

uchchwhash committed:
structured timeslices; no juxtaposing yet

uchchwhash committed:
functionally complete?

uchchwhash committed:
remove explicit dependence on xr_apply

uchchwhash committed:
Simply interface

- drop `validate_construction` method
- replace `Drop` with a construction time `dataset_filter`
  for `BasicProduct`s
